src/SIL.Harmony/ResourceService.cs

```diff
@@ -0,0 +1,162 @@
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Options;
+using SIL.Harmony.Changes;
+using SIL.Harmony.Core;
+using SIL.Harmony.Db;
+using SIL.Harmony.Helpers;
+using SIL.Harmony.Resource;
+
+namespace SIL.Harmony;
+
+public class ResourceService
+{
+    private readonly CrdtRepository _crdtRepository;
+    private readonly IOptions<CrdtConfig> _crdtConfig;
+    private readonly DataModel _dataModel;
+
+    internal ResourceService(CrdtRepository crdtRepository, IOptions<CrdtConfig> crdtConfig, DataModel dataModel)
+    {
+        _crdtRepository = crdtRepository;
+        _crdtConfig = crdtConfig;
+        _dataModel = dataModel;
+    }
+
+    private void ValidateResourcesSetup()
+    {
+        if (!_crdtConfig.Value.RemoteResourcesEnabled) throw new RemoteResourceNotEnabledException();
+    }
+
+    public async Task<CrdtResource> AddLocalResource(string resourcePath,
+        Guid clientId,
+        Guid id = default,
+        IRemoteResourceService? resourceService = null)
+    {
+        ValidateResourcesSetup();
+        var localResource = new LocalResource
+        {
+            Id = id == default ? Guid.NewGuid() : id,
+            LocalPath = Path.GetFullPath(resourcePath)
+        };
+        if (!localResource.FileExists()) throw new FileNotFoundException(localResource.LocalPath);
+        await using var transaction = await _crdtRepository.BeginTransactionAsync();
+        await _crdtRepository.AddLocalResource(localResource);
+        if (resourceService is not null)
+        {
+            var uploadResult = await resourceService.UploadResource(localResource.LocalPath);
```

Comment:
Currently we roll everything back if the upload fails.
But, the local resource is still valid.
I think it would make more sense to follow the same path we would if resourceService was `null`. I.e. keep the local resource and add a `CreateRemoteResourcePendingUploadChange`